Additional tutorials (#110)
SKopecz/PositiveIntegrators.jl: docs/src/stratospheric_reaction.md

There are two independent linear invariants, e.g. ``u_1+u_2+3u_3+2u_4+u_5+2u_6=(1,1,3,2,1,2)\cdot\mathbf{u}_0`` and ``u_5+u_6 = 1.097⋅10^9``.

The stratospheric reaction problem can be represented as a (non-conservative) PDS with production terms
```math
\begin{aligned}
p_{13} &= r_5, & p_{21} &= r_6, & p_{22} &= r_1+r_{10},\\
p_{23} &= r_3, & p_{24} &= r_1,& p_{32} &= r_2,\\
p_{41} &= r_7, & p_{42}&= r_4+r_9, & p_{43}&= r_4+r_7+r_8,\\
p_{44} &= r_3+r_5, & p_{56}&=r_9+r_{10}, & p_{65}&=r_8+r_{11}.
\end{aligned}
```
and additional destruction terms
```math
\begin{aligned}
d_{22}&= r_{11}, & d_{44}&=r_2.
\end{aligned}
```

## Solution of the production-destruction system

Now we are ready to define a `PDSProblem` and to solve this problem with a method of [PositiveIntegrators.jl](https://github.com/SKopecz/PositiveIntegrators.jl) or [OrdinaryDiffEq.jl](https://docs.sciml.ai/OrdinaryDiffEq/stable/). 

As mentioned above, we will try two approaches to solve this PDS and compare their efficiency. These are
1. an out-of-place implementation with standard (dynamic) matrices and vectors,

> 2. an out-of-place implementation with static matrices and vectors from [StaticArrays.jl](https://juliaarrays.github.io/StaticArrays.jl/stable/).

Author: Added.

    r7 = k7 * O1D * O3
    r8 = k8 * O3 * NO
    r9 = k9 * NO2 * O
    r10 = k10 * NO2
    r11 = k11 * NO * O

    return [0.0 0.0 r5 0.0 0.0 0.0;
            r6 r1+r10 r3 r1 0.0 0.0;
            0.0 r2 0.0 0.0 0.0 0.0;
            r7 r4+r9 r4+r7+r8 r3+r5 0.0 0.0;
            0.0 0.0 0.0 0.0 0.0 r9+r10;
            0.0 0.0 0.0 0.0 r8+r11 0.0]
end

function dest(u, p, t)
    O1D, O, O3, O2, NO, NO2 = u

    k2 = 8.018e-17
    k11 = 1.0e-8

    r2 = k2 * O * O2
    r11 = k11 * NO * O

    return [0.0, r11, 0.0, r2, 0.0, 0.0]
end
nothing #hide
```
The solution of the stratospheric reaction problem can now be computed as follows.
```@example stratreac
u0 = [9.906e1, 6.624e8, 5.326e11, 1.697e16, 4e6, 1.093e9] # initial values
tspan = (4.32e4, 3.024e5) # time domain
prob = PDSProblem(prod, dest, u0, tspan) # create the PDS

sol = solve(prob, MPRK43I(1.0, 0.5))

nothing #hide
```
```@example stratreac
using Plots

plot(sol,
    layout=(3,2),
    xguide = "t [h]",
    xguidefontsize = 8,
    xticks = (range(first(tspan), last(tspan), 4), range(12.0, 84.0, 4)),
    yguide=["O¹ᴰ" "O" "O₃" "O₂" "NO" "NO₂"],
    tickfontsize = 7,
    legend = :none, 
    widen = true
    )
```

> As MPRK schemes do not preserve general linear invariants, especially when applied to non-conservative PDS, we compute and plot the relative errors with respect to both linear invariants to see how well these are preserved.

Reviewer: Shall we also show some results with methods from OrdinaryDiffEq.jl?

Author: I thought about it, but thought this would fit better in the "Benchmarks" section suggested in #7. Do you agree?

Reviewer: It would be nice to include this in some benchmarks, too 👍 However, I think it would still be nice to have a qualitative comparison here with one standard method that is able to conserve all linear invariants.

Author: Added.

# [Tutorial: Solution of a stratospheric reaction problem](@id tutorial-stratos)

This tutorial is about the efficient solution of a stiff non-autonomous and non-conservative production-destruction systems (PDS) with a small number of differential equations. 
We will compare the use of standard arrays and static arrays from [StaticArrays.jl](https://juliaarrays.github.io/StaticArrays.jl/stable/) and assess their efficiency.

## Definition of the production-destruction system

> This stratospheric reaction problem was described by Adrian Sandu in [Positive Numerical Integration Methods for Chemical Kinetic Systems](https://doi.org/10.1006/jcph.2001.6750), see also the paper [Positivity-preserving adaptive Runge–Kutta methods](https://doi.org/10.2140/camcos.2021.16.155) by Stefan Nüßlein, Hendrik Ranocha and David I. Ketcheson. The goverining equations are

Reviewer: ```suggestion
This stratospheric reaction problem was described by Adrian Sandu in [Positive Numerical Integration Methods for Chemical Kinetic Systems](https://doi.org/10.1006/jcph.2001.6750), see also the paper [Positivity-preserving adaptive Runge–Kutta methods](https://doi.org/10.2140/camcos.2021.16.155) by Stefan Nüßlein, Hendrik Ranocha and David I. Ketcheson. The governing equations are
```

There are two independent linear invariants, e.g. ``u_1+u_2+3u_3+2u_4+u_5+2u_6=(1,1,3,2,1,2)\cdot\mathbf{u}_0`` and ``u_5+u_6 = 1.097⋅10^9``.

The stratospheric reaction problem can be represented as a (non-conservative) PDS with production terms
```math
\begin{aligned}
p_{13} &= r_5, & p_{21} &= r_6, & p_{22} &= r_1+r_{10},\\
p_{23} &= r_3, & p_{24} &= r_1,& p_{32} &= r_2,\\
p_{41} &= r_7, & p_{42}&= r_4+r_9, & p_{43}&= r_4+r_7+r_8,\\
p_{44} &= r_3+r_5, & p_{56}&=r_9+r_{10}, & p_{65}&=r_8+r_{11}.
\end{aligned}
```
and additional destruction terms
```math
\begin{aligned}
d_{22}&= r_{11}, & d_{44}&=r_2.
\end{aligned}
```

## Solution of the production-destruction system

Now we are ready to define a `PDSProblem` and to solve this problem with a method of [PositiveIntegrators.jl](https://github.com/SKopecz/PositiveIntegrators.jl) or [OrdinaryDiffEq.jl](https://docs.sciml.ai/OrdinaryDiffEq/stable/). 

As mentioned above, we will try different approaches to solve this PDS and compare their efficiency. These are
1. an out-of-place implementation with standard (dynamic) matrices and vectors,

> 2. an in-place implementation with standard (dynamics) matrices and vectors,

Reviewer: ```suggestion
2. an in-place implementation with standard (dynamic) matrices and vectors,
```

### Preservation of linear invariants
As MPRK schemes do not preserve general linear invariants, especially when applied to non-conservative PDS, we compute and plot the relative errors with respect to both linear invariants to see how well these are preserved.

```@example stratreac
linear_invariant(a, u) = sum(a .* u)

function relerr_lininv(a, u0, sol)
    c = linear_invariant(a, u0)
    return abs.(c .- (x -> linear_invariant(a, x)).(sol.u))./c
end

a1 = [1; 1; 3; 2; 1; 2] # first linear invariant
a2 = [0; 0; 0; 0; 1; 1] # second linear invariant

p1 = plot(sol_oop.t, relerr_lininv(a1, u0, sol_oop))
p2 = plot(sol_oop.t, relerr_lininv(a2, u0, sol_oop))
plot(p1, p2, 
    xticks = (range(first(tspan), last(tspan), 4), range(12.0, 84.0, 4)),
    legend = :none)
```

In contrast to MPRK schemes, Runge-Kutta and Rosenbrock methods preserve all linear invariants. We show this using the Rosenbrock scheme `Rosenbrock23` as an example, making sure that the solution is actually nonnegative.

```@example stratreac
using OrdinaryDiffEq

> 
> sol_Ros23 = solve(prob_oop, Rosenbrock23(),
>              isoutofdomain = (y, p, t) -> any(x -> x < 0, y) # reject step if the current solution has a negative element
>              );

Reviewer: ```suggestion

# reject step if the current solution has a negative element
isoutofdomain(u, args...) = any(<(0), u) # p, t are additional arguments 
sol_Ros23 = solve(prob_oop, Rosenbrock23(); isoutofdomain);
```

Author: I'd suggest to take this even further by providing a function `isnegative` (and maybe also `isnonnegative`). If a solution is negative, I'd also like to know which components are affected and “how negative” the solution actually is.

Here's a draft. 

```julia
julia> using PositiveIntegrators, OrdinaryDiffEq

julia> function isnegative(u::AbstractVector, args...)
           return any(<(0), u)
       end
isnegative (generic function with 1 method)

julia> function isnegative(u::AbstractVector{<:AbstractVector}, args...)
           anyisnegative = any(isnegative, u)
           if anyisnegative
               components = Int64[]
               minima = eltype(first(u))[]
               # look for components with negative elements
               for i in eachindex(first(u))
                   v = getindex.(u,i)
                   if isnegative(v)
                       push!(components, i)
                       push!(minima, minimum(v))            
                   end
               end
               println("Components ", components ," contain negative elements, the respective minima are ", minima)
           end    
           return anyisnegative
       end
isnegative (generic function with 2 methods)

julia> isnonnegative(args...) = !isnegative(args...)
isnonnegative (generic function with 1 method)

julia> sol = solve(prob_pds_stratreac, Rosenbrock23());

julia> isnegative(sol.u)
Components [1, 2, 5] contain negative elements, the respective minima are [-1.5591986975545963e-23, -9.712921995482734e-9, -9.170872661434066e-9]
true

julia> isnonnegative(sol.u)
Components [1, 2, 5] contain negative elements, the respective minima are [-1.5591986975545963e-23, -9.712921995482734e-9, -9.170872661434066e-9]
false

julia> sol2 = solve(prob_pds_stratreac, Rosenbrock23(); isoutofdomain = isnegative);

julia> isnegative(sol2.u)
false

julia> isnonnegative(sol2.u)
true

```

I think we should check in all tutorials that the MPRK solutions are actually nonnegative, since this is what the package promises.

Reviewer: That's a nice idea

> any(map(x -> any(x .< 0), sol_Ros23.u)) # Is there a negative solution component?

Author: Is this the way to check if the solution is nonnegative?

Reviewer: Yes, that works. But you can also use the functional form like

```suggestion
any(isoutofdomain, sol_Ros23.u) # Is there a negative solution component?
```

### Performance comparison

Finally, we use [BenchmarkTools.jl](https://github.com/JuliaCI/BenchmarkTools.jl)

> to show to compare the different implementations and to show the benefit of using static arrays.

Reviewer: ```suggestion
to compare the different implementations and to show the benefit of using static arrays.
```
